libraries/core/src/topics.rs

```diff
@@ -94,3 +86,172 @@ impl Display for DataflowId {
         }
     }
 }
+
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
+pub struct DataflowResult {
+    pub uuid: Uuid,
+    pub timestamp: uhlc::Timestamp,
+    pub node_results: BTreeMap<NodeId, Result<(), NodeError>>,
+}
+
+impl DataflowResult {
+    pub fn ok_empty(uuid: Uuid, timestamp: uhlc::Timestamp) -> Self {
+        Self {
+            uuid,
+            timestamp,
+            node_results: Default::default(),
+        }
+    }
+
+    pub fn is_ok(&self) -> bool {
+        self.node_results.values().all(|r| r.is_ok())
+    }
+
+    pub fn root_error(&self) -> RootError<'_> {
+        RootError(self)
+    }
+}
+
+pub struct RootError<'a>(&'a DataflowResult);
+
+impl std::fmt::Display for RootError<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let failed = self
+            .0
+            .node_results
+            .iter()
+            .filter_map(|(id, r)| r.as_ref().err().map(|e| (id, e)));
+        let total_failed = failed.clone().count();
+
+        let mut non_cascading: Vec<_> = failed
+            .clone()
+            .filter(|(_, e)| !matches!(e.cause, NodeErrorCause::Cascading { .. }))
+            .collect();
+        non_cascading.sort_by_key(|(_, e)| e.timestamp);
+        // try to print earliest non-cascading error
+        if let Some((id, err)) = non_cascading.first() {
+            write!(f, "Node `{id}` failed: {err}")?;
+        } else {
+            // no non-cascading errors -> print earliest cascading
+            let mut all: Vec<_> = failed.collect();
+            all.sort_by_key(|(_, e)| e.timestamp);
+            if let Some((id, err)) = all.first() {
+                write!(f, "Node `{id}` failed: {err}")?;
+            } else {
+                write!(f, "unknown error")?;
+            }
+        }
+
+        if total_failed > 1 {
+            write!(
+                f,
+                "\n\nThere are {} more errors. Check the `out/{}` folder for full details.",
```

Comment:
True. I slightly adjusted the message in 450002e to clarify that only consequential errors are hidden (i.e. errors that only occurred because of the shown errors)